Source/Core/InputCommon/ControllerInterface/Device.cpp

```diff
@@ -301,27 +390,59 @@ DeviceContainer::DetectInput(u32 wait_ms, const std::vector<std::string>& device
   if (device_states.empty())
     return {};
 
-  u32 time = 0;
-  while (time < wait_ms)
+  std::vector<InputDetection> detections;
+
+  const auto start_time = Clock::now();
+  while (true)
   {
+    const auto now = Clock::now();
+    const auto elapsed_time = now - start_time;
+
+    if (elapsed_time >= maximum_wait || (detections.empty() && elapsed_time >= initial_wait) ||
+        (!detections.empty() && detections.back().release_time.has_value() &&
+         now >= *detections.back().release_time + confirmation_wait))
+    {
+      break;
+    }
+
     Common::SleepCurrentThread(10);
-    time += 10;
 
     for (auto& device_state : device_states)
     {
-      for (auto& input_state : device_state.input_states)
+      for (std::size_t i = 0; i != device_state.input_states.size(); ++i)
       {
-        // We want an input that was initially 0.0 and currently 1.0.
-        const auto detection_score =
-            (input_state.input.GetState() - std::abs(input_state.initial_state));
-
-        if (detection_score > INPUT_DETECT_THRESHOLD)
-          return {device_state.device, &input_state.input};
+        auto& input_state = device_state.input_states[i];
+        input_state.Update();
+
+        if (input_state.IsPressed())
+        {
+          input_state.is_ready = false;
+
+          // Digital presses will evaluate as 1 here.
+          // Analog presses will evaluate greater than 1.
+          const auto smoothness =
```

Comment:
Do you foresee a need for this 'smoothness' value or would a boolean / enumeration dictating whether the key is analog or digital be clearer?

Reply:
(I saw you you already checked for 'digital' keys earlier)

Reply:
I guess I left it as `smoothness` instead of an `is_analog` boolean because the check isn't foolproof. If you quickly pull your trigger so it's never seen as half-pressed then `smoothness` will be 1.
Also I might tweak input detect logic a bit more relating to `SPURIOUS_TRIGGER_COMBO_THRESHOLD`. If users are finding their button combos not mapping because of that logic, then `smoothness` could be increased to require a slow-pull to consider a combo "spurious".